File: src/eslint.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/wait.h>
#include <unistd.h>

#include "eslint.h"
#include "util.h"
#include "workspace.h"

void eslint_init(void) {
    log_info("ESLint subsystem initialized");
}

char *eslint_find_binary(const char *file_path) {
    if (!file_path) {
        return str_dup("eslint");
    }
    
    char *workspace_root = workspace_get_root_for_file(file_path);
    
    if (!workspace_root) {
        return str_dup("eslint");
    }
    
    char *binary = workspace_find_eslint_binary(workspace_root);

    free(workspace_root);
    
    return binary;
}

char *eslint_find_workspace_root(const char *file_path) {
    return workspace_get_root_for_file(file_path);
}

static char *run_eslint_command(const char *eslint_bin, const char *file_path, 
                                 const char *working_dir, bool fix, const char *content) {
    int stdin_pipe[2];
    int stdout_pipe[2];

    if (pipe(stdin_pipe) == -1 || pipe(stdout_pipe) == -1) {
        log_error("Failed to create pipes");

        return NULL;
    }

    pid_t pid = fork();

    if (pid == -1) {
        log_error("Failed to fork process");

        close(stdin_pipe[0]);
        close(stdin_pipe[1]);
        close(stdout_pipe[0]);
        close(stdout_pipe[1]);

        return NULL;
    }

    if (pid == 0) {
        close(stdin_pipe[1]);
        close(stdout_pipe[0]);
        dup2(stdin_pipe[0], STDIN_FILENO);
        dup2(stdout_pipe[1], STDOUT_FILENO);
        dup2(stdout_pipe[1], STDERR_FILENO);
        close(stdin_pipe[0]);
        close(stdout_pipe[1]);

        if (working_dir) {
            chdir(working_dir);
        }

        if (content) {
            char stdin_filename_arg[4096];
            snprintf(stdin_filename_arg, sizeof(stdin_filename_arg), 
                     "--stdin-filename=%s", file_path);

            if (fix) {
                execlp(eslint_bin, eslint_bin, "--format", "json", "--fix", 
                       "--stdin", stdin_filename_arg, NULL);
            } else {
                execlp(eslint_bin, eslint_bin, "--format", "json", 
                       "--stdin", stdin_filename_arg, NULL);
            }
        } else {
            if (fix) {
                execlp(eslint_bin, eslint_bin, "--format", "json", "--fix", file_path, NULL);
            } else {
                execlp(eslint_bin, eslint_bin, "--format", "json", file_path, NULL);
            }
        }

        exit(1);
    }

    close(stdin_pipe[0]);
    close(stdout_pipe[1]);

    if (content) {
        size_t content_len = strlen(content);
        ssize_t written = write(stdin_pipe[1], content, content_len);

        if (written != (ssize_t)content_len) {
            log_error("Failed to write all content to stdin");
        }
    }

    close(stdin_pipe[1]);

    size_t buffer_size = 4096;
    size_t total_size = 0;
    char *output = malloc(buffer_size);

    if (!output) {
        close(stdout_pipe[0]);

        return NULL;
    }

    ssize_t bytes_read;

    while ((bytes_read = read(stdout_pipe[0], output + total_size, buffer_size - total_size - 1)) > 0) {
        total_size += bytes_read;

        if (total_size + 1 >= buffer_size) {
            buffer_size *= 2;
            char *new_output = realloc(output, buffer_size);

            if (!new_output) {
                free(output);
                close(stdout_pipe[0]);

                return NULL;
            }

            output = new_output;
        }
    }

    output[total_size] = '\0';
    close(stdout_pipe[0]);
    int status;
    waitpid(pid, &status, 0);

    return output;
}

static eslint_result_t *parse_eslint_output(const char *json_output, const char *file_path) {
    if (!json_output || !file_path) {
        return NULL;
    }
    
    json_object *root = json_tokener_parse(json_output);
    if (!root) {
        log_error("Failed to parse ESLint JSON output");

        return NULL;
    }
    
    if (!json_object_is_type(root, json_type_array)) {
        json_object_put(root);

        return NULL;
    }
    
    eslint_result_t *result = calloc(1, sizeof(eslint_result_t));
    if (!result) {
        json_object_put(root);

        return NULL;
    }
    
    result->file_path = str_dup(file_path);
    
    // ESLint returns an array with one element per file
    if (json_object_array_length(root) == 0) {
        json_object_put(root);

        return result;
    }
    
    json_object *file_result = json_object_array_get_idx(root, 0);
    json_object *messages_obj;
    
    if (!json_object_object_get_ex(file_result, "messages", &messages_obj)) {
        json_object_put(root);

        return result;
    }
    
    int message_count = json_object_array_length(messages_obj);

    if (message_count == 0) {
        json_object_put(root);

        return result;
    }
    
    result->diagnostics = calloc(message_count, sizeof(eslint_diagnostic_t));
    result->diagnostic_count = message_count;
    
    for (int i = 0; i < message_count; i++) {
        json_object *msg = json_object_array_get_idx(messages_obj, i);
        eslint_diagnostic_t *diag = &result->diagnostics[i];
        
        json_object *line_obj, *column_obj, *end_line_obj, *end_column_obj;
        json_object *severity_obj, *message_obj, *rule_id_obj, *fix_obj;
        
        if (json_object_object_get_ex(msg, "line", &line_obj)) {
            diag->line = json_object_get_int(line_obj);
        }
        
        if (json_object_object_get_ex(msg, "column", &column_obj)) {
            diag->column = json_object_get_int(column_obj);
        }
        
        if (json_object_object_get_ex(msg, "endLine", &end_line_obj)) {
            diag->end_line = json_object_get_int(end_line_obj);
        } else {
            diag->end_line = diag->line;
        }
        
        if (json_object_object_get_ex(msg, "endColumn", &end_column_obj)) {
            diag->end_column = json_object_get_int(end_column_obj);
        } else {
            diag->end_column = diag->column;
        }
        
        if (json_object_object_get_ex(msg, "severity", &severity_obj)) {
            diag->severity = json_object_get_int(severity_obj);
        }
        
        if (json_object_object_get_ex(msg, "message", &message_obj)) {
            diag->message = str_dup(json_object_get_string(message_obj));
        }
        
        if (json_object_object_get_ex(msg, "ruleId", &rule_id_obj)) {
            diag->rule_id = str_dup(json_object_get_string(rule_id_obj));
        }
        
        if (json_object_object_get_ex(msg, "fix", &fix_obj)) {
            diag->fixable = json_object_get_boolean(fix_obj);
        }
    }
    
    json_object_put(root);

    return result;
}

eslint_result_t *eslint_lint_file(const char *file_path, const char *content) {
    if (!file_path) {
        return NULL;
    }

    char *eslint_bin = eslint_find_binary(file_path);
    char *workspace_root = eslint_find_workspace_root(file_path);
    char *package_root = workspace_find_package_root(file_path);
    char *working_dir = package_root ? package_root : workspace_root;

    log_debug("Running ESLint: %s on %s (working dir: %s, stdin: %s)", 
              eslint_bin, file_path, working_dir ? working_dir : "none",
              content ? "yes" : "no");

    char *output = run_eslint_command(eslint_bin, file_path, working_dir, false, content);

    free(eslint_bin);
    free(workspace_root);

    if (package_root) {
        free(package_root);
    }

    if (!output) {
        log_error("ESLint command failed");

        return NULL;
    }

    eslint_result_t *result = parse_eslint_output(output, file_path);

    free(output);

    return result;
}

eslint_result_t *eslint_fix_file(const char *file_path) {
    if (!file_path) {
        return NULL;
    }

    char *eslint_bin = eslint_find_binary(file_path);
    char *workspace_root = eslint_find_workspace_root(file_path);
    char *package_root = workspace_find_package_root(file_path);
    char *working_dir = package_root ? package_root : workspace_root;

    log_debug("Running ESLint fix: %s on %s", eslint_bin, file_path);

    char *output = run_eslint_command(eslint_bin, file_path, working_dir, true, NULL);

    free(eslint_bin);
    free(workspace_root);

    if (package_root) {
        free(package_root);
    }

    if (!output) {
        log_error("ESLint fix command failed");

        return NULL;
    }

    eslint_result_t *result = parse_eslint_output(output, file_path);

    free(output);

    return result;
}

void eslint_free_result(eslint_result_t *result) {
    if (!result) {
        return;
    }
    
    if (result->file_path) {
        free(result->file_path);
    }
    
    if (result->diagnostics) {
        for (int i = 0; i < result->diagnostic_count; i++) {
            if (result->diagnostics[i].message) {
                free(result->diagnostics[i].message);
            }
            if (result->diagnostics[i].rule_id) {
                free(result->diagnostics[i].rule_id);
            }
        }
        free(result->diagnostics);
    }
    
    if (result->fixed_output) {
        free(result->fixed_output);
    }
    
    free(result);
}

File: src/Eslint.zig
const std = @import("std");
const Workspace = @import("Workspace.zig");

const Eslint = @This();

pub const Diagnostic = struct {
    line: i32,
    column: i32,
    end_line: i32,
    end_column: i32,
    severity: i32, // 1=warning, 2=error (ESLint convention)
    message: []const u8,
    rule_id: ?[]const u8,
    fixable: bool,
};

pub const LintResult = struct {
    allocator: std.mem.Allocator,
    file_path: []const u8,
    diagnostics: []Diagnostic,
    fixed_output: ?[]const u8,

    pub fn deinit(self: *LintResult) void {
        for (self.diagnostics) |*d| {
            self.allocator.free(d.message);
            if (d.rule_id) |r| self.allocator.free(r);
        }
        self.allocator.free(self.diagnostics);
        self.allocator.free(self.file_path);
        if (self.fixed_output) |o| self.allocator.free(o);
    }
};

allocator: std.mem.Allocator,
workspace: *Workspace,

pub fn init(allocator: std.mem.Allocator, workspace: *Workspace) Eslint {
    return .{
        .allocator = allocator,
        .workspace = workspace,
    };
}

/// Lint a file, optionally with in-memory content (unsaved buffer).
/// Returns null if ESLint produces no output or the binary is not found.
/// Caller must call LintResult.deinit() on the returned value.
pub fn lintFile(self: *Eslint, file_path: []const u8, content: ?[]const u8) !?LintResult {
    const eslint_bin = try self.findBinary(file_path);
    defer self.allocator.free(eslint_bin);

    const workspace_root = try self.workspace.getRootForFile(file_path);
    defer if (workspace_root) |r| self.allocator.free(r);

    const package_root = try self.workspace.findPackageRoot(file_path);
    defer if (package_root) |r| self.allocator.free(r);

    const cwd = package_root orelse workspace_root;

    std.log.debug("Running ESLint: {s} on {s} (working dir: {s}, stdin: {})", .{
        eslint_bin,
        file_path,
        cwd orelse "(none)",
        content != null,
    });

    const json_output = try self.runCommand(.{
        .eslint_bin = eslint_bin,
        .file_path = file_path,
        .cwd = cwd,
        .fix = false,
        .content = content,
    }) orelse return null;
    defer self.allocator.free(json_output);

    return self.parseOutput(json_output, file_path);
}

/// Run ESLint with --fix on a file. Returns the result with fixed_output if fixes applied.
/// Caller must call LintResult.deinit() on the returned value.
pub fn fixFile(self: *Eslint, file_path: []const u8) !?LintResult {
    const eslint_bin = try self.findBinary(file_path);
    defer self.allocator.free(eslint_bin);

    const workspace_root = try self.workspace.getRootForFile(file_path);
    defer if (workspace_root) |r| self.allocator.free(r);

    const package_root = try self.workspace.findPackageRoot(file_path);
    defer if (package_root) |r| self.allocator.free(r);

    const cwd = package_root orelse workspace_root;

    std.log.debug("Running ESLint fix: {s} on {s}", .{ eslint_bin, file_path });

    const json_output = try self.runCommand(.{
        .eslint_bin = eslint_bin,
        .file_path = file_path,
        .cwd = cwd,
        .fix = true,
        .content = null,
    }) orelse return null;
    defer self.allocator.free(json_output);

    return self.parseOutput(json_output, file_path);
}

// ---- Private ----

fn findBinary(self: *Eslint, file_path: []const u8) ![]const u8 {
    const root = try self.workspace.getRootForFile(file_path);
    defer if (root) |r| self.allocator.free(r);
    const workspace_root = root orelse return self.allocator.dupe(u8, "eslint");
    return self.workspace.findEslintBinary(workspace_root);
}

const RunArgs = struct {
    eslint_bin: []const u8,
    file_path: []const u8,
    cwd: ?[]const u8,
    fix: bool,
    content: ?[]const u8,
};

/// Spawns ESLint as a child process and returns its stdout.
/// Returns null on spawn failure. Caller owns the returned slice.
fn runCommand(self: *Eslint, args: RunArgs) !?[]const u8 {
    var argv: std.ArrayList([]const u8) = .empty;
    defer argv.deinit(self.allocator);

    try argv.append(self.allocator, args.eslint_bin);
    try argv.append(self.allocator, "--format");
    try argv.append(self.allocator, "json");
    if (args.fix) try argv.append(self.allocator, "--fix");

    // stdin-filename flag — must outlive the argv list
    var stdin_flag_buf: [std.fs.max_path_bytes + 32]u8 = undefined;
    if (args.content != null) {
        const stdin_flag = try std.fmt.bufPrint(
            &stdin_flag_buf,
            "--stdin-filename={s}",
            .{args.file_path},
        );
        try argv.append(self.allocator, "--stdin");
        try argv.append(self.allocator, stdin_flag);
    } else {
        try argv.append(self.allocator, args.file_path);
    }

    var child = std.process.Child.init(argv.items, self.allocator);
    child.stdin_behavior = if (args.content != null) .Pipe else .Ignore;
    child.stdout_behavior = .Pipe;
    child.stderr_behavior = .Ignore;
    if (args.cwd) |cwd| child.cwd = cwd;

    child.spawn() catch |err| {
        std.log.err("Failed to spawn ESLint: {}", .{err});
        return null;
    };

    // Write content to stdin if provided
    if (args.content) |content| {
        child.stdin.?.writeAll(content) catch {};
        child.stdin.?.close();
        child.stdin = null;
    }

    // Read all stdout
    var out_buf: [1024 * 1024]u8 = undefined; // 1 MiB read buffer
    var reader = child.stdout.?.readerStreaming(&out_buf);
    const stdout = reader.interface.allocRemaining(self.allocator, .unlimited) catch |err| {
        _ = child.wait() catch {};
        std.log.err("Failed to read ESLint output: {}", .{err});
        return null;
    };

    _ = child.wait() catch {};
    return stdout;
}

// ESLint JSON output types for typed deserialization
const EslintFix = struct {
    range: [2]i32 = .{ 0, 0 },
    text: []const u8 = "",
};

const EslintMessage = struct {
    line: i32 = 0,
    column: i32 = 0,
    endLine: ?i32 = null,
    endColumn: ?i32 = null,
    severity: i32 = 0,
    message: []const u8 = "",
    ruleId: ?[]const u8 = null,
    fix: ?EslintFix = null,
};

const EslintFileResult = struct {
    filePath: []const u8 = "",
    messages: []EslintMessage = &.{},
    output: ?[]const u8 = null,
};

/// Parse ESLint `--format json` output into a LintResult.
/// Returns null if the JSON is invalid or empty.
fn parseOutput(self: *Eslint, json_bytes: []const u8, file_path: []const u8) !?LintResult {
    // ESLint may emit non-JSON preamble (startup messages, warnings).
    // Find the first '[' to skip any prefix noise.
    const json_start = std.mem.indexOfScalar(u8, json_bytes, '[') orelse {
        std.log.warn("ESLint output contains no JSON array", .{});
        return null;
    };

    const parsed = std.json.parseFromSlice(
        []EslintFileResult,
        self.allocator,
        json_bytes[json_start..],
        .{ .ignore_unknown_fields = true },
    ) catch |err| {
        std.log.warn("Failed to parse ESLint JSON output: {}", .{err});
        return null;
    };
    defer parsed.deinit();

    if (parsed.value.len == 0) {
        return LintResult{
            .allocator = self.allocator,
            .file_path = try self.allocator.dupe(u8, file_path),
            .diagnostics = &.{},
            .fixed_output = null,
        };
    }

    const file_result = parsed.value[0];

    // Convert EslintMessage[] -> Diagnostic[]
    var diags = try self.allocator.alloc(Diagnostic, file_result.messages.len);
    errdefer self.allocator.free(diags);

    var diag_count: usize = 0;
    errdefer {
        for (diags[0..diag_count]) |*d| {
            self.allocator.free(d.message);
            if (d.rule_id) |r| self.allocator.free(r);
        }
    }

    for (file_result.messages, 0..) |msg, i| {
        diags[i] = .{
            .line = msg.line,
            .column = msg.column,
            .end_line = msg.endLine orelse msg.line,
            .end_column = msg.endColumn orelse msg.column,
            .severity = msg.severity,
            .message = try self.allocator.dupe(u8, msg.message),
            .rule_id = if (msg.ruleId) |r| try self.allocator.dupe(u8, r) else null,
            .fixable = msg.fix != null,
        };
        diag_count += 1;
    }

    const fixed_output = if (file_result.output) |o| try self.allocator.dupe(u8, o) else null;

    return LintResult{
        .allocator = self.allocator,
        .file_path = try self.allocator.dupe(u8, file_path),
        .diagnostics = diags,
        .fixed_output = fixed_output,
    };
}

// ---- Tests ----

test "parseOutput parses ESLint JSON output" {
    const Workspace_ = @import("Workspace.zig");
    var fake_workspace = Workspace_{
        .allocator = std.testing.allocator,
        .root_path = null,
        .is_pnpm_workspace = false,
    };
    var eslint = Eslint.init(std.testing.allocator, &fake_workspace);

    const json =
        \\[{"filePath":"/test.ts","messages":[{"line":1,"column":5,"endLine":1,"endColumn":10,"severity":2,"message":"no-unused-vars","ruleId":"no-unused-vars","fix":{"range":[4,9],"text":""}}]}]
    ;

    var result = (try eslint.parseOutput(json, "/test.ts")).?;
    defer result.deinit();

    try std.testing.expectEqualStrings("/test.ts", result.file_path);
    try std.testing.expectEqual(@as(usize, 1), result.diagnostics.len);

    const d = result.diagnostics[0];
    try std.testing.expectEqual(@as(i32, 1), d.line);
    try std.testing.expectEqual(@as(i32, 5), d.column);
    try std.testing.expectEqual(@as(i32, 1), d.end_line);
    try std.testing.expectEqual(@as(i32, 10), d.end_column);
    try std.testing.expectEqual(@as(i32, 2), d.severity);
    try std.testing.expectEqualStrings("no-unused-vars", d.message);
    try std.testing.expectEqualStrings("no-unused-vars", d.rule_id.?);
    try std.testing.expect(d.fixable);
}

test "parseOutput handles empty messages array" {
    const Workspace_ = @import("Workspace.zig");
    var fake_workspace = Workspace_{
        .allocator = std.testing.allocator,
        .root_path = null,
        .is_pnpm_workspace = false,
    };
    var eslint = Eslint.init(std.testing.allocator, &fake_workspace);

    const json =
        \\[{"filePath":"/test.ts","messages":[]}]
    ;

    var result = (try eslint.parseOutput(json, "/test.ts")).?;
    defer result.deinit();

    try std.testing.expectEqual(@as(usize, 0), result.diagnostics.len);
}

test "parseOutput handles missing optional fields" {
    const Workspace_ = @import("Workspace.zig");
    var fake_workspace = Workspace_{
        .allocator = std.testing.allocator,
        .root_path = null,
        .is_pnpm_workspace = false,
    };
    var eslint = Eslint.init(std.testing.allocator, &fake_workspace);

    const json =
        \\[{"filePath":"/test.ts","messages":[{"line":3,"column":1,"severity":1,"message":"some warning"}]}]
    ;

    var result = (try eslint.parseOutput(json, "/test.ts")).?;
    defer result.deinit();

    const d = result.diagnostics[0];
    try std.testing.expectEqual(@as(i32, 3), d.end_line); // defaults to line
    try std.testing.expectEqual(@as(i32, 1), d.end_column); // defaults to column
    try std.testing.expect(d.rule_id == null);
    try std.testing.expect(!d.fixable);
}

test "parseOutput returns null on invalid JSON" {
    const Workspace_ = @import("Workspace.zig");
    var fake_workspace = Workspace_{
        .allocator = std.testing.allocator,
        .root_path = null,
        .is_pnpm_workspace = false,
    };
    var eslint = Eslint.init(std.testing.allocator, &fake_workspace);

    const result = try eslint.parseOutput("not json at all", "/test.ts");
    try std.testing.expect(result == null);
}

test "parseOutput skips non-JSON prefix" {
    const Workspace_ = @import("Workspace.zig");
    var fake_workspace = Workspace_{
        .allocator = std.testing.allocator,
        .root_path = null,
        .is_pnpm_workspace = false,
    };
    var eslint = Eslint.init(std.testing.allocator, &fake_workspace);

    const json = "some startup noise\n[{\"filePath\":\"/t.ts\",\"messages\":[]}]";

    var result = (try eslint.parseOutput(json, "/t.ts")).?;
    defer result.deinit();

    try std.testing.expectEqual(@as(usize, 0), result.diagnostics.len);
}

test "severity mapping: ESLint 2=error, 1=warning" {
    const Workspace_ = @import("Workspace.zig");
    var fake_workspace = Workspace_{
        .allocator = std.testing.allocator,
        .root_path = null,
        .is_pnpm_workspace = false,
    };
    var eslint = Eslint.init(std.testing.allocator, &fake_workspace);

    const json =
        \\[{"filePath":"/t.ts","messages":[{"line":1,"column":1,"severity":2,"message":"err"},{"line":2,"column":1,"severity":1,"message":"warn"}]}]
    ;

    var result = (try eslint.parseOutput(json, "/t.ts")).?;
    defer result.deinit();

    try std.testing.expectEqual(@as(i32, 2), result.diagnostics[0].severity);
    try std.testing.expectEqual(@as(i32, 1), result.diagnostics[1].severity);
}
